### DISCUSSION TITLE
Ability to modify head from non layout pages

### DISCUSSION THREAD

**Discussion by dearcodes** [withastro/roadmap#360] [Proposal]
The `head` component usually has common tags (usually defined in the layout page) and page-specific tags for SEO purposes (among others). This is common practice among modern web applications. 

Seems like currently there's only one way to define a common head for all pages. 

Are there any plans to include a mechanism to modify head tags from page-level Astro components?

**Comment by delucis** [comment]
Astro’s SSR implementation streams your page’s HTML to the browser as it processes, allowing `<head>` to be sent even before the rest of the page’s template has been handled on the server. This helps speed up performance for users, but means any changes to `<head>` _must_ all happen inside the `<head>` element itself.

The pattern we usually recommend is for your layout to receive props to customize the `<head>`.

**Reply by smnbbrv** [reply]
@delucis is there a possibility to disable this feature (or at least wait in header until all the content is rendered)?

**Example 1**

I build websites with static caching which means I can guarantee that in 99.9% of the cases they are returned to the user directly from CDN without hitting Astro backend. 

This means, this feature (that is of course a performance-perfect default scenario)  is actually playing against me in this particular case. I don't get any performance bonus because the way astro gives its page to CDN / Cache service does not anyhow matter; meanwhile cannot push `<Head>` logic from the components, that would save me from the code clutter and give a lot of freedom.

It would be great to have two different scenarios:

- current header handling in Layouts (default)
- [a next.js-alike header handling](https://nextjs.org/docs/api-reference/next/head) where the page is waiting for being rendered, but enables features like `astro/head`. This should be a must-opt-in for the project developer, with a clear explanation what are the downsides. For this I assume the following is required:
  - some setting in astro.config.js, that disables default rendering flow / buffers the whole HTML before sending to the user
  - `astro/head` library that provides head tags elevation, works only if the buffering is on

---

The page buffering would also be useful for those who for some reasons wants to modify HTML output after it was produced.

**Example 2**

I want to allow syntax e.g. like `#variableName#` in my CMS and across my codebase and then I would like to replace all those variables with real values before serving it to the end-user or CDN. Currently it is only possible to intercept the variable values by processing CMS data, for the codebase I would need to use a different logic, which in the end results in two implementations of the same thing. Or, there should be another server on top of this one that replaces the values.

This particular example is not common and probably cannot alone be a good reason to implement such a thing, however it shows that there are use-cases that require to delay the response until the whole page is rendered.

**Reply by delucis** [reply]
My guess is that this could create a pretty big fork in rendering logic that might be hard to maintain, but I’m not familiar with the internals here, so will leave it to someone else to say whether this is viable.

**Reply by smnbbrv** [reply]
@delucis I can't say for the whole project and all relevant adapters, but it seems pretty straitforward. I use node middleware SSR and this is a relevant piece of code that gets generated:

```js
// entry.mjs
async function writeWebResponse(app, res, webResponse) {
  const { status, headers, body } = webResponse;
  if (app.setCookieHeaders) {
    const setCookieHeaders = Array.from(app.setCookieHeaders(webResponse));
    if (setCookieHeaders.length) {
      res.setHeader("Set-Cookie", setCookieHeaders);
    }
  }
  res.writeHead(status, Object.fromEntries(headers.entries()));
  if (body) {
    for await (const chunk of body) {
      res.write(chunk);
    }
  }
  res.end();
}
```

If we just naively adapt these lines

```js
  if (body) {
    for await (const chunk of body) {
      res.write(chunk);
    }
  }
  res.end();
```

to 

```js
  let response = '';

  if (body) {
    for await (const chunk of body) {
      response += chunk;
    }
  }

  // and here we can mix <head> items in

  res.send(chunk);
```

that would already do the magic.

**Reply by jamesarosen** [reply]
Another use-case: I have a component that will use resources from a third-party domain. That means I need to modify `Content-Security-Policy` to add the relevant `script-src` and other directives on any page where that component is used. With Astro's current behavior, that means I have two choices:

 1. set `Content-Security-Policy` in my Layout to the union of all the possible sources I use anywhere in the app. This is less secure.
 2. have individual pages pass a `Content-Security-Policy` to the Layout. This means that anyone developing a page needs to know all of the possible connection locations of all of the components they use. More importantly, it means if someone removes a component, they may leave behind vestigial resource locations.

Similarly, if this component loads JS or a font from a third-party, I probably want to add a `<link rel="prefetch">` for it.

**Reply by matthewp** [reply]
I have a new proposal to enable this: https://github.com/withastro/roadmap/discussions/481
